File: android/app/src/main/java/com/vixel/studio/core/model/Transition.kt
package com.vixel.studio.core.model

enum class TransitionType(val label: String) {
    NONE("None"),
    DISSOLVE("Dissolve"),
    FADE_BLACK("Fade to black"),
    FADE_WHITE("Fade to white"),
    SLIDE_LEFT("Slide left"),
    SLIDE_RIGHT("Slide right"),
    SLIDE_UP("Slide up"),
    SLIDE_DOWN("Slide down"),
    WIPE_LEFT("Wipe left"),
    WIPE_RIGHT("Wipe right"),
    ZOOM_IN("Zoom in"),
    ZOOM_OUT("Zoom out"),
    BLUR("Blur"),
    ;

    val isNone: Boolean get() = this == NONE
}

/**
 * A transition *into* a clip, overlapping it with the one before.
 *
 * Because the two clips overlap, a transition shortens the timeline by its own
 * duration — which is why [Project.durationUs] subtracts it rather than just
 * summing clip lengths.
 */
data class Transition(
    val type: TransitionType = TransitionType.NONE,
    val durationUs: Long = 500_000L,
) {
    val isActive: Boolean get() = !type.isNone && durationUs > 0L

    companion object {
        val NONE = Transition(TransitionType.NONE, 0L)
        const val MIN_US = 100_000L
        const val MAX_US = 2_000_000L
    }
}

/**
 * What the compositor should draw at one instant.
 *
 * Outside a transition only [primaryIndex] is set. Inside one, [fromIndex] is
 * the outgoing clip and [progress] runs 0 to 1 across the overlap.
 */
data class Composition(
    val primaryIndex: Int,
    val fromIndex: Int = -1,
    val progress: Float = 1f,
    val transition: Transition = Transition.NONE,
) {
    val isTransitioning: Boolean get() = fromIndex >= 0 && transition.isActive
}

File: flutter/lib/core/model/transition.dart
enum TransitionType {
  none('None', ''),
  dissolve('Dissolve', 'fade'),
  fadeBlack('Fade to black', 'fadeblack'),
  fadeWhite('Fade to white', 'fadewhite'),
  slideLeft('Slide left', 'slideleft'),
  slideRight('Slide right', 'slideright'),
  slideUp('Slide up', 'slideup'),
  slideDown('Slide down', 'slidedown'),
  wipeLeft('Wipe left', 'wipeleft'),
  wipeRight('Wipe right', 'wiperight'),
  circleOpen('Circle open', 'circleopen'),
  circleClose('Circle close', 'circleclose'),
  zoomIn('Zoom in', 'zoomin'),
  dissolveBlur('Blur', 'fadegrays');

  const TransitionType(this.label, this.xfadeName);

  final String label;

  /// Name of the matching FFmpeg xfade transition.
  ///
  /// Keeping the mapping on the enum is what stops the preview and the export
  /// drifting: adding a transition means filling in both columns here, and a
  /// missing xfade name is visible at the definition rather than discovered
  /// when a render comes out wrong.
  final String xfadeName;

  bool get isNone => this == TransitionType.none;
}

class Transition {
  const Transition({this.type = TransitionType.none, this.durationUs = 500000});

  final TransitionType type;
  final int durationUs;

  bool get isActive => !type.isNone && durationUs > 0;

  static const Transition none =
      Transition(type: TransitionType.none, durationUs: 0);

  static const int minUs = 100000;
  static const int maxUs = 2000000;

  Transition copyWith({TransitionType? type, int? durationUs}) => Transition(
        type: type ?? this.type,
        durationUs: durationUs ?? this.durationUs,
      );

  Map<String, dynamic> toJson() => {'type': type.name, 'durationUs': durationUs};

  static Transition fromJson(Map<String, dynamic> json) => Transition(
        type: TransitionType.values.firstWhere(
          (t) => t.name == json['type'],
          orElse: () => TransitionType.none,
        ),
        durationUs: (json['durationUs'] as num?)?.toInt() ?? 500000,
      );

  @override
  bool operator ==(Object other) =>
      other is Transition &&
      other.type == type &&
      other.durationUs == durationUs;

  @override
  int get hashCode => Object.hash(type, durationUs);
}
